typescript/infra/scripts/verify-validator.ts

```diff
@@ -0,0 +1,297 @@
+import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
+import yargs from 'yargs';
+
+const MAX_MISSING_CHECKPOINTS = 10;
+
+interface Checkpoint {
+  checkpoint: {
+    outbox_domain: number;
+    root: string;
+    index: number;
+  };
+  signature: {
+    r: string;
+    s: string;
+    v: number;
+  };
+}
+
+function isCheckpoint(obj: unknown): obj is Checkpoint {
+  const c = obj as Partial<Checkpoint>;
+  return (
+    typeof obj == 'object' &&
+    obj != null &&
+    'checkpoint' in obj &&
+    Number.isSafeInteger(c.checkpoint?.outbox_domain) &&
+    Number.isSafeInteger(c.checkpoint?.index) &&
+    isValidHashStr(c.checkpoint?.root ?? '') &&
+    'signature' in obj &&
+    isValidHashStr(c.signature?.r ?? '') &&
+    isValidHashStr(c.signature?.s ?? '') &&
+    Number.isSafeInteger(c.signature?.v)
+  );
+}
+
+function isValidHashStr(s: string): boolean {
+  return !!s.match(/^0x[0-9a-f]{1,64}$/im);
+}
+
+function getArgs() {
+  return yargs(process.argv.slice(2))
+    .alias('a', 'address')
+    .describe('a', 'address of the validator to inspect')
+    .demandOption('a')
+    .string('a')
+    .alias('p', 'prospective')
+    .describe('p', 'S3 bucket of the prospective validator')
+    .demandOption('p')
+    .string('p')
+    .alias('c', 'control')
+    .describe('c', 'S3 bucket of the the known (control) validator')
+    .demandOption('c')
+    .string('c').argv;
+}
+
+class S3Wrapper {
+  private readonly client: S3Client;
+  readonly region: string;
+  readonly bucket: string;
+
+  constructor(bucketUrl: string) {
+    const match = bucketUrl.match(
+      /^(?:https?:\/\/)?(.*)\.s3\.(.*)\.amazonaws.com\/?$/,
+    );
+    if (!match) throw new Error('Could not parse bucket url');
+    this.bucket = match[1];
+    this.region = match[2];
+    this.client = new S3Client({ region: this.region });
+  }
+
+  async getS3Obj<T = unknown>(
+    key: string,
+  ): Promise<{ obj: T; modified: Date }> {
+    const response = await this.client.send(
+      new GetObjectCommand({
+        Bucket: this.bucket,
+        Key: key,
+      }),
+    );
+    if (!response.Body) {
+      throw new Error('No data received');
+    }
+    const bodyStream: NodeJS.ReadableStream =
+      'stream' in response.Body
+        ? response.Body.stream()
+        : (response.Body as NodeJS.ReadableStream);
+
+    const body: string = await streamToString(bodyStream);
+    return {
+      obj: JSON.parse(body),
+      modified: response.LastModified!,
+    };
+  }
+}
+
+async function main() {
+  const {
+    a: _validatorAddress,
+    p: prospectiveBucket,
+    c: controlBucket,
+  } = await getArgs();
+
+  const cClient = new S3Wrapper(controlBucket);
```

Comment:
how would you feel about expanding `c` to `control` and `p` to `prospective` here & elsewhere? Just noticing that the single letter differentiators are easy to mistake